ts/keys.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2014 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Matt Tracy (matt.r.tracy@gmail.com)
+
+package ts
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/storage/engine"
+	"github.com/cockroachdb/cockroach/util/encoding"
+)
+
+// Time series keys are carefully constructed to usefully sort the data in the
+// key-value store for queries. Time series data is typically queryied by metric
+// name and a timespan; we therefore expose the metric name and the timestamp
+// prominently on the key.
+//
+// The precise formula for a time series key is:
+//   [system key prefix]tsd[metric name][resolution][timestamp][source key]
+//
+// The metric name is an arbitrary string identifying the metric, although the
+// ts system may enforce naming rules at a higher level. This string is binary
+// encoded in the key.
+//
+// The resolution refers to the sample duration at which data is stored. A
+// perfect resolution sample would keep the original timestamp and value for
+// every data point; lower resolution samples aggregate data points which fall
+// within the same sample period, keeping only the sum and count of all data
+// points. Multiple resolutions may be kept for a single metric; lower
+// resolutions are used to quickly query larger timespans, while higher
+// resolutions are used to provide more detail for smaller timespans. The
+// resolution is denoted in the key by a three character string.
+//
+// The timestamp is the first timestamp for which this key might contain data.
```

Comment:
I found this wording a bit confusing: "The timestamp is the first timestamp" and "this key". I imagine that you're rounding the timestamp to the resolution. 

This long comment is helpful, but you could make it better by providing an example or two of encoding a metric name and timestamp at 2 different resolutions.